### ISSUE TITLE
The use of regex named capture groups is not consistent

### ISSUE THREAD

**Issue by RG255** [PowerShell/PowerShell#12000]
# Steps to reproduce

In the attached script
[Regex.zip](https://github.com/PowerShell/PowerShell/files/4288316/Regex.zip)
There are two filters 'get-dns' and 'get-dns1'
There is also a regular expression 'RegexDNSEntry'
This regex contains a number of capture groups (many are ignored) and in particular one that has a name 'DNS'
Depending on how the regex is used (see the comments at the start of each filter) in one you have to know the numerical offset of the group you want to use, in the other you can actually user the name of the capture group e.g. 'DNS'
e.g:
$Length = $Result.Matches.groups[1].length (get-dns)
$Length = $Result.Groups['DNS'].length (get-dns1)
both methods work but one - using the name, seems to be the most programmatically safe because if the regex is altered and the group order and hence the numerical offset changes in the first example breaking the code. When looking at the $Result.Matches.groups (both examples) output, the name is known but I have so far not been able to work out how to *easily* to access the group by name it is only accessible numerically.

There is data within the script that can be used to demonstrate the workings and you will see the output from both is the same.
```

The behaviour should be the same however the regex is used and the output/group should be accessible by either name or numerical offset

```

See the code and examples

```

# Environment data

<!-- provide the output of $PSVersionTable -->
Name                           Value
----                           -----
PSVersion                      7.0.0-rc.3
PSEdition                      Core
GitCommitId                    7.0.0-rc.3
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

**Comment by SeeminglyScience** [comment]
The `Matches` property is an array of matches.  When you then call `Groups` on the result of that property expression, the `Groups` property of `Match` is called for each item in the array, similar to `(Get-ChildItem).Name`.  The property enumeration creates a `object[]` (an array of any type) which does not have the indexer you're trying to access.

tl;dr, do this instead:

```powershell
$result.Matches.ForEach{ $PSItem.Groups['DNS'] }.Length
# or for groups of a single match
$result.Matches[0].Groups['DNS'].Length
```

**Comment by RG255** [comment]
@SeeminglyScience 

OK thanks for that I see that it works, however it would be better if it was consistent across both methods:
This is the option for the first example
[DBG]: PS C:\>> $result.matches
Groups   : {0, DNS}
Success  : True
Name     : 0
Captures : {0}
Index    : 0
Length   : 26
Value    : 0.0.0.0     	 fr.a2dfp.net

And this the second
[DBG]: PS C:\>> $result
Groups   : {0, DNS}
Success  : True
Name     : 0
Captures : {0}
Index    : 0
Length   : 26
Value    : 0.0.0.0     	 fr.a2dfp.net

As you can see the output of $Result is the same but the method of access is different I agree there is a way to gather data from the named group using the first method but as the subject says the method is not consistent. 

Using Measure-Command the time taken is considerable less using the second method:

(Measure-Command {$Result = (Select-String -InputObject $InputObject -Pattern $RegexDNSEntry -AllMatches)}).Ticks 
177220
6097
1801
5095
26834
2157
1888
1893
2004
2090
2004
2026
5723
1917
1915
1928
2008
1858
1844
1897
1829
1815
1790
1984
1830

(Measure-Command ({$Result = $RegexDNSEntry.Match($InputObject)})).Ticks
76175
433
295
411
350
303
274
325
353
319
306
354
344
357
361
347
590
352
356
390
332
383
322
385
399

This difference is the same magnitude in both V5.1 and V7RC3

If you think this is a non issue then feel free to close the case.

**Comment by SeeminglyScience** [comment]
> As you can see the output of $Result is the same but the method of access is different I agree there is a way to gather data from the named group using the first method but as the subject says the method is not consistent.

It is in the same way that `1` has the same output as `1, 2, 3`.  When an object that can be enumerated (e.g. `array`/`object[]`/`ArrayList`/`List<>`) is outputted to the pipeline (in this case implicitly to the host), the object is enumerated and items contents are displayed instead.  You can check this with the `GetType()` method that is present on all objects:

```powershell
$result.GetType()

# vs

$result.Matches.GetType()
```

It *is* technically consistent with how the language works, but it is admittedly pretty confusing. I absolutely remember getting tripped up on this early on (which is why I jumped on the issue) and have seen the same confusion in others many times.  This API in particular makes that interaction makes it really easy to trip on.  Not sure what would be done differently though, or if that's just gotta be one of those things that folks run into.

> Using Measure-Command the time taken is considerable less using the second method:

That's pretty much always going to be the case.  Commands have a lot more overhead involved than a simple method invocation.  Honestly less than 1ms difference is pretty good, I'd have expected the difference to be more drastic.
